File: recipes/java-recipes/spring-boot-postgres-recipe/src/main/java/com/trendyol/stove/examples/java/spring/application/external/category/CategoryHttpApiImpl.java
package com.trendyol.stove.examples.java.spring.application.external.category;

import com.trendyol.stove.recipes.shared.application.category.CategoryApiResponse;

import org.springframework.web.reactive.function.client.WebClient;

import reactor.core.publisher.Mono;

public class CategoryHttpApiImpl implements CategoryHttpApi {
    private final WebClient categoryWebClient;

    public CategoryHttpApiImpl(WebClient categoryWebClient) {
        this.categoryWebClient = categoryWebClient;
    }

    @Override
    public Mono<CategoryApiResponse> getCategoryById(int id) {
        return categoryWebClient
                .get()
                .uri("/categories/{id}", id)
                .retrieve()
                .bodyToMono(CategoryApiResponse.class);
    }
}

File: recipes/kotlin-recipes/ktor-recipe/src/main/kotlin/com/trendyol/stove/examples/kotlin/ktor/application/external/CategoryHttpApiImpl.kt
package com.trendyol.stove.examples.kotlin.ktor.application.external

import com.trendyol.stove.recipes.shared.application.category.CategoryApiResponse
import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.plugins.*
import io.ktor.client.request.*
import io.ktor.http.*
import kotlin.time.Duration.Companion.seconds

class CategoryHttpApiImpl(
  private val httpClient: HttpClient,
  private val categoryApiConfiguration: CategoryApiConfiguration
) : CategoryHttpApi {
  override suspend fun getCategory(id: Int): CategoryApiResponse {
    return httpClient
      .get("${categoryApiConfiguration.url}/categories/$id") {
        accept(ContentType.Application.Json)
        timeout {
          requestTimeoutMillis = categoryApiConfiguration.timeout.seconds.inWholeMilliseconds
          connectTimeoutMillis = categoryApiConfiguration.timeout.seconds.inWholeMilliseconds
          socketTimeoutMillis = categoryApiConfiguration.timeout.seconds.inWholeMilliseconds
        }
      }
      .body()
  }
}
